Stream CDM CloudWatch logs to S3 with SNS/SQS notifications

dav3r opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR sets up a process to stream the data in the [CDM CloudWatch log groups](https://github.com/cisagov/cool-sharedservices-cdm/blob/45710e672e52966348d7642250efdda3e319360d/variables.tf#L71-L75) to an S3 bucket so that it can be ingested by the CDM system.  It includes an SNS topic that receives notifications when new objects are added to the S3 bucket, as well as an SQS queue that is subscribed to the SNS topic.  The SNS topic and SQS queues are [required by the tool](https://docs.splunk.com/Documentation/AddOns/released/AWS/SQS-basedS3) that CDM plans to use for ingesting this data.

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

The tool used by CDM is not able to assume a role to ingest the CloudWatch logs directly from CloudWatch.  However it is able to assume a role to ingest the data if it is in an S3 bucket with an SNS topic and SQS queue as defined in this PR.  The data flow looks like this:

```mermaid
flowchart LR
    A(CloudWatch Log Groups with subscription filters) -->|Amazon Data Firehose| B(S3 Bucket)
    B -->|S3 Notifications| C(SNS Topic)
    C -->|Subscription| D(SQS Queue)
    C -->|Undeliverable Messages| E[SQS Dead Letter Queue]
```

Note that this PR does **not** include the permissions changes required for the CDM user to access the S3 bucket, SNS topic, and SQS queue - that will be handled in a separate PR (currently in #45, but subject to change).

> [!NOTE]
>  ~The `blocked` tag has been added to this PR since it is now dependent on https://github.com/cisagov/cool-cdm-cloudtrail-tf-module/pull/31.  This note and the `blocked` tag will be removed when https://github.com/cisagov/cool-cdm-cloudtrail-tf-module/pull/31 is merged.~

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

I successfully applied these changes with no errors in the Staging Shared Services account, then I confirmed all of the following:
 * Requests are flowing through the Firehose stream and being delivered to the S3 bucket (visible on the "Monitoring" tab of the Firehose web console)
 * Public access to the bucket is disabled (visible on the "Permissions" tab of the S3 web console)
 * The S3 bucket is encrypted with server-side encryption with Amazon S3 managed keys (SSE-S3) (visible on the "Properties" tab of the S3 web console)
 * Objects are in the S3 bucket and being sorted in folders by date and hour (visible in the "Objects" tab of the S3 web console)
 * The S3 objects contain correct message data from the CDM log groups/streams (verified by downloading a few and looking at their content)
 * The S3 bucket is correctly configured to notify the SNS topic for "All object create events" (visible on the "Properties" tab of the S3 web console)
 * The SNS topic is deployed with the correct access policy allowing only the S3 bucket to publish to the topic (visible in the "Access policy" tab of the SNS web console)
 * The SQS queue is subscribed to the SNS topic and the subscription is confirmed (visible in the "Subscriptions" tab of the SNS web console as well as in the "SNS subscriptions" tab of the SQS web console)
 * Messages are continuously arriving in the SQS queue, awaiting pickup by the CDM tool (visible in the "Monitoring" tab of the SQS web console)
 * The SQS queue is using the correct SQS dead-letter queue (visible in the "Dead-letter queue" tab of the SQS web console)
 * Both SQS queues (main and dead-letter) are encrypted with server-side encryption, managed by SSE SQS (visible in the "Encryption" tab of the SQS web console)
 * Both SQS queues (main and dead-letter) are deployed with the correct access policy allowing only the SNS topic to send messages to the queue (visible in the "Access policy" tab of the SQS web console)

<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

## ✅ Pre-merge checklist ##
- [x] Revert testing branch commit https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/a079104e492868faa1a4b120d64a62f65fc5890e

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [ ] Apply these changes in Production
- [ ] Work with CDM engineers to ensure that the CDM role has the correct permissions to successfully fetch the CloudWatch data from the S3 bucket (will be handled as part of #45)

dav3r committed:
Add variables related to exporting CloudWatch logs for CDM

dav3r committed:
Define S3 bucket to store CDM CloudWatch logs

dav3r committed:
Define the SNS topic that receives notifications from the S3 bucket and is subscribed to by the SQS queue

dav3r committed:
Configure the bucket to send notifications to the SNS topic whenever objects are added to the bucket

dav3r committed:
Define SQS queues to receive notifications from the SNS topic

dav3r committed:
Define policy and role that allow CloudWatch to send log events to a Firehose delivery stream

dav3r committed:
Define policy and role that allow Firehose to send data to the CDM S3 bucket

dav3r committed:
Define Firehose delivery stream and CloudWatch log subscription filters

This lets us automatically send CloudWatch log events to the Firehose stream which then pushes them to our bucket.

dav3r committed:
Add new outputs related to CloudWatch log export for CDM

dav3r committed:
Update all Terraform docs in the README

jsf9k reviewed (approved):
Nice work!  The diagram is a nice touch.

dav3r committed:
Set SQS queue visibility timeout to 10 minutes

When using the default value of 30 seconds, Splunk gives an error stating: "It is recommended to set VisibilityTimeout larger than 5 minutes for the specified SQS queue."

dav3r committed:
Use version 5.x of the Terraform AWS provider

Version 5.x is required in order to use the "Decompression" processor in the aws_kinesis_firehose_delivery_stream resource (part of the extended_s3_configuration).

dav3r committed:
Temporarily use a testing branch of cisagov/cool-cdm-cloudtrail-tf-module

That branch is also using version 5.x of the AWS provider.

This commit will be reverted after https://github.com/cisagov/cool-cdm-cloudtrail-tf-module/pull/31 is merged.

dav3r committed:
Decompress each item that passes through the Firehose delivery stream

The CloudWatch subscription filter automatically GZIPs the log events before sending them to the Firehose delivery stream, but we want the data to be decompressed before it is written to the S3 bucket so that Splunk can ingest it properly.  For more information, see: https://docs.aws.amazon.com/firehose/latest/dev/writing-with-cloudwatch-logs.html

dav3r committed:
Update README with latest output from terraform-docs

jsf9k reviewed (approved):
It's so nice I approved it twice!

mcdonnnj commented on README.md:
Please update the repository description in the README to mention this pretty significant additional functionality.

dav3r replied:
Please have a look at https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/354b667092685a3e88178bca9f8ac281901c9eff.

dav3r committed:
Revert "Temporarily use a testing branch of cisagov/cool-cdm-cloudtrail-tf-module"

This reverts commit a079104e492868faa1a4b120d64a62f65fc5890e.

mcdonnnj commented on cloudwatch_logs_for_cdm.tf:
```diff
@@ -0,0 +1,52 @@
+  for_each = toset(var.cloudwatch_policy_log_groups)
+
+  destination_arn = aws_kinesis_firehose_delivery_stream.cloudwatch_logs.arn
+  filter_pattern  = ""
```
Might be worth mentioning in a comment that this matches everything.

dav3r replied:
Good thinking - I expanded the comment in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/762a214980a3dc92e3ac993a683ada9d3a91c585.

mcdonnnj commented on cloudwatch_to_firehose_role.tf:
```diff
@@ -0,0 +1,68 @@
+    effect = "Allow"
+
+    resources = [
+      "arn:aws:firehose:${var.aws_region}:${local.sharedservices_account_id}:deliverystream/${var.firehose_delivery_stream_name}",
```
Are we unable to use `aws_kinesis_firehose_delivery_stream.cloudwatch_logs.arn`? I prefer not to handcraft ARNs unless specifically necessary.

dav3r replied:
We cannot due to a circular dependency.  I added an explanatory comment in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/c3748705bf3573ee9af6b9ec58ed4c4c8f8a13ed.

mcdonnnj commented on cloudwatch_sns.tf:
```diff
@@ -0,0 +1,51 @@
+  }
+}
+
+# Attach the policy (that allows the CDM S3 bucket to send notifications to the
+# SNS topic) to the SNS topic.
```
I get why they're here but can this be reworded to remove the parentheses?

dav3r replied:
Try https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/c4edbc67d4dad15f8b322bba74f0ca6c97150924 on for size and let me know what you think.

mcdonnnj replied:
Seems like a good fit!

mcdonnnj commented on variables.tf:
```diff
@@ -56,6 +61,18 @@ variable "aws_region" {
   default     = "us-east-1"
 }
 
+variable "cloudwatch_logs_sqs_queue_name" {
```
It might be worth mentioning that this is also used as a base for the dead letter queue's name.

dav3r replied:
Good idea.  I did that in https://github.com/cisagov/cool-sharedservices-cdm/pull/46/commits/8df490c3610a1e649c0dcd48d72c0422445064f2.

mcdonnnj reviewed (approved):
This is looking pretty solid. Strong work 💪💪💪 I have some minor requests/suggestions for your consideration.

dav3r committed:
Update README description to reflect additional functionality

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve a comment to be more explanatory

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Improve readability of a comment

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add additional information to a variable description

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

dav3r committed:
Add a comment explaining why an ARN is being constructed manually

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
